File: py/morphe_surface/resolve.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, cast

from .refs import resolve_ref, schema_type, unwrap_nullable

if TYPE_CHECKING:
    from .hints import MorpheHint
    from .strategies import Strategy

_SCALAR_TYPES = {"string", "integer", "number", "boolean"}


def resolve_strategy(
    schema: dict[str, Any], hint: MorpheHint, *, root: dict[str, Any] | None = None
) -> Strategy:
    """Single chokepoint (ADR-0014 D3): (schema shape, hint) -> closed Strategy.

    Hints win over structural inference; structure is the hint-free floor. An
    unrecognized construct degrades to ``diagnostic-node`` (totality, D8) — never raises.
    ``root`` lets array items behind a local ``$ref`` participate in inference.
    """
    if hint.strategy is not None:
        return hint.strategy
    schema = unwrap_nullable(schema, root or {})
    if "enum" in schema:
        return "badge"
    t = schema_type(schema)
    if t == "object":
        return "record-card"
    if t == "array":
        items = _resolved(schema.get("items"), root)
        # Only a FLAT record row lowers to a grid (KRA-640); nested or shapeless rows
        # keep the per-row disclosure path.
        return "table" if items is not None and _flat_record(items, root) else "card-stack"
    if t in _SCALAR_TYPES:
        return "scalar"
    return "diagnostic-node"


def _flat_record(items: dict[str, Any], root: dict[str, Any] | None) -> bool:
    if schema_type(items) != "object":
        return False
    props = items.get("properties")
    if not isinstance(props, dict) or not props:
        return False
    return all(
        (sub := _resolved(raw, root)) is not None and schema_type(sub) not in {"object", "array"}
        for raw in props.values()
    )


def _resolved(schema: object, root: dict[str, Any] | None) -> dict[str, Any] | None:
    if not isinstance(schema, dict):
        return None
    return resolve_ref(cast("dict[str, Any]", schema), root or {})

File: src/lib/surface-edge/resolve.ts
import type { MorpheHint } from "./hints.js";
import { isSchema, resolveRef, schemaType, unwrapNullable } from "./refs.js";
import type { JsonSchema, Strategy } from "./spec.js";

const SCALAR_TYPES = new Set(["string", "integer", "number", "boolean"]);

/** The sole `(schema shape, hint) -> closed strategy` decision point. */
export function resolveStrategy(
	schema: JsonSchema,
	hint: MorpheHint,
	root: JsonSchema = {},
): Strategy {
	if (hint.strategy !== undefined) return hint.strategy;
	const resolved = unwrapNullable(schema, root);
	if (Object.hasOwn(resolved, "enum")) return "badge";
	const type = schemaType(resolved);
	if (type === "object") return "record-card";
	if (type === "array") {
		const rawItems = resolved.items;
		const items = isSchema(rawItems) ? resolveRef(rawItems, root) : null;
		return items !== null && flatRecord(items, root) ? "table" : "card-stack";
	}
	if (type !== null && SCALAR_TYPES.has(type)) return "scalar";
	return "diagnostic-node";
}

function flatRecord(items: JsonSchema, root: JsonSchema): boolean {
	if (schemaType(items) !== "object" || !isSchema(items.properties)) return false;
	const properties = Object.values(items.properties);
	if (properties.length === 0) return false;
	return properties.every((raw) => {
		if (!isSchema(raw)) return false;
		const type = schemaType(resolveRef(raw, root));
		return type !== "object" && type !== "array";
	});
}
